File: CivitasRuby/pkg/CivitasRuby-0.0.1/lib/casilla.rb
#encoding:utf-8

require_relative 'jugador'
require_relative 'tipo_casilla'
require_relative 'titulo_propiedad'
require_relative 'tipo_casilla'
require_relative 'sorpresa'

module Civitas
  
    attr_reader :titulo , :nombre
  
    class Casilla
      private_class_method :new
      attr_reader :nombre ,:titulo
      attr_accessor :nombre


      @@Carcel = 3
      
      def self.Carcel
        @@Carcel
      end
      
      def initialize(nombre , importe , titulo , mazo , sorpresa , tipo)

        @nombre = nombre
        @importe = importe
        @titulo = titulo    
        @mazo = mazo
        @sorpresa = sorpresa
        @tipo = tipo
        

      end

      def self.new_por_nombre(nombre)
        tipo = TipoCasilla::DESCANSO
        
        if(nombre == "Salida")
          tipo = TipoCasilla::DESCANSO
        elsif(nombre == "Cárcel")
          tipo = TipoCasilla::DESCANSO
        elsif(nombre == "Juez")
          tipo = TipoCasilla::JUEZ
        elsif(nombre == "Aparcamiento")
          tipo = TipoCasilla::CALLE
        else
          tipo = TipoCasilla::Calle
        end
        
        new(nombre,0,nil,nil,nil,tipo)

      end

      def self.new_por_titulo(titulo)
        tipo = TipoCasilla::CALLE
        new(titulo.nombre,titulo.precio_compra,titulo,nil,nil,tipo)
        
      end
      
      def self.new_por_cantidad(cantidad,nombre)
        tipo = TipoCasilla::IMPUESTO
        
        new(nombre,cantidad,nil,nil,nil,tipo)
      end

      def self.new_juez(numCasillaCarcel)
        tipo = TipoCasilla::JUEZ
        new("Juez",0,nil,nil,nil,tipo)
      end
      
      def self.new_sorpresa(mazo,nombre)
        tipo = TipoCasilla::SORPRESA
        
        new("Sorpresa",0,nil,nil,nil,tipo)
      end
      


      def to_string
        
        resultado = "Casilla: "+@nombre+"\nimporte:"+@importe.to_s+"\n\n"
        
        if(@titulo != nil)
          resultado+=@titulo.to_string+"\n\n"
        end
        
        
        
        resultado
        
        
      end


      #METODOS PRIVADOS   
      private
      
      def informe(iactual,todos)
        Diario.instance.ocurre_evento(todos[iactual].nombre + " ha caido en #{@nombre}")

      end

      def recibe_jugador_impuesto(iactual , todos)

        if( jugador_correcto(iactual,todos) )
         
          informe(iactual,todos)
          todos[iactual].paga_impuesto(@importe)
        end
      end

      def recibe_jugador_calle(iactual , todos)
        if(jugador_correcto(iactual,todos))
            informe(iactual,todos)
            jugador=todos[iactual]

            if(@titulo.tiene_propietario and !@titulo.propietario.encarcelado)
               
              if(@titulo.propietario != jugador and ! @titulo.hipotecado)
                jugador.puede_comprar = false
                jugador.paga_alquiler(@titulo.get_precio_alquiler)
                @titulo.propietario.recibe(@titulo.get_precio_alquiler)
                
              elsif(@titulo.hipotecado)
                jugador.puede_comprar = false
              end
            
            else
              
              jugador.puede_comprar = true
            end
        end
      end

      public
      def recibe_jugador(iactual,todos)
          case @tipo
          when TipoCasilla::CALLE
              recibe_jugador_calle(iactual,todos)
          when TipoCasilla::IMPUESTO
              recibe_jugador_impuesto(iactual,todos)
          when TipoCasilla::JUEZ
              recibe_jugador_juez(iactual,todos)
          when TipoCasilla::SORPRESA
              recibe_jugador_sorpresa(iactual,todos)
          else
              informe(iactual,todos)
          end
      end
      
      private
      def jugador_correcto(iactual,todos)

        iactual >= 0 && todos.size() > iactual

      end

      def recibe_jugador_juez(iactual,todos)
        if(jugadorCorrecto(iactual, todos))
          informe(iactual, todos);
          todos[iactual].encarcelar(@@carcel);
        end
      end

      def recibe_jugador_sorpresa(iactual , todos)
        if(jugador_correcto(iactual,todos))
          informe(iactual,todos)
          @sorpresa=@mazo.siguiente
          @sorpresa.aplicar_a_jugador(iactual, todos)
        end
      end
  end
end

File: CivitasJava/src/civitas/Casilla.java
package civitas;

import java.util.ArrayList;

public class Casilla{




    private static int carcel; //numero de la casilla de carcel
    
    private String nombre;
    private float importe;
    
    //ASOCIACIONES
    private TipoCasilla tipo;
    private TituloPropiedad tituloPropiedad;
    private Sorpresa sorpresa;
    private MazoSorpresas mazo;
    
    
    
    
    Casilla(String nombre)
    {
        init();
        tipo = TipoCasilla.DESCANSO;
        this.nombre = nombre;
        
    }
    
    
    
    
    Casilla(TituloPropiedad titulo)
    {
        init();
        tipo = TipoCasilla.CALLE;
        tituloPropiedad=titulo;
        nombre = titulo.getNombre();
        
    }
    
    
    
    
    Casilla(float cantidad , String nombre){
        init();
        this.importe=cantidad;
        this.nombre=nombre;
        this.tipo = TipoCasilla.IMPUESTO;
    }
    
    
    
    Casilla(int numCasillaCarcel , String nombre){
        init();
        this.carcel=numCasillaCarcel;
        this.nombre=nombre;
        this.tipo = TipoCasilla.JUEZ;
    }
    
    
    
    Casilla(MazoSorpresas mazo , String nombre){
        init();
        this.mazo=mazo;
        this.nombre=nombre;
        tipo = TipoCasilla.SORPRESA;
    }
    
    

    
    public static int getCarcel()
    {
        return carcel;
    }   
    
     
     
    void recibeJugador(int iactual , ArrayList<Jugador> todos)
    {
        switch(tipo){
            case CALLE:
                recibeJugador_calle(iactual,todos);
                break;
            case IMPUESTO:
                recibeJugador_impuesto(iactual,todos);
                break;
            case JUEZ:
                recibeJugador_juez(iactual,todos);
                break;
            case SORPRESA:
                recibeJugador_sorpresa(iactual,todos);
                break;
            default:
                informe(iactual,todos);
                        
                    }
        
    }
    
    public String getNombre(){
        return nombre;
    }
        
    TituloPropiedad getTituloPropiedad()
    {
        return tituloPropiedad;
    }
    
    @Override //Sobrecarga del método de la clase padre Object 
    public String toString(){
        
        String info = "Casilla :"+nombre
                +"\n\timporte: "+importe
                +"\n\tMazo:"+mazo
                +"\n\tSorpresa:"+sorpresa
                +"\n\ttipo:"+tipo
                +"\n\tTitulo de propiedad:"+tituloPropiedad
                +"\n\tCárcel:"+carcel;
        
        return info;
        
    }
    
    
    public boolean jugadorCorrecto(int iactual , ArrayList<Jugador> todos){
        
        boolean indiceCorrecto = iactual >= 0 && todos.size() > iactual;
        
        return indiceCorrecto;
    }
    
    private void informe(int iactual , ArrayList<Jugador> todos){
        Diario.getInstance().ocurreEvento("El jugador "+todos.get(iactual).getNombre()
        +" ha caido en la casilla "+nombre+toString());
    }
    
    private void init(){
        nombre = "";
        tituloPropiedad = null;
        sorpresa = null;
        mazo = null;
        importe = 0f;
            
    }
    
    
    private void recibeJugador_calle(int actual , ArrayList<Jugador> todos)
    {
        if(jugadorCorrecto(actual,todos))
        {
            informe(actual,todos);
            Jugador jugador=todos.get(actual);
            
            if(!tituloPropiedad.tienePropietario())
            {
                jugador.puedeComprarCasilla();
            }
            else
            {
                tituloPropiedad.tramitarAlquiler(jugador);
                
            }
             
        }
        
    }
    
    private void recibeJugador_impuesto(int iactual , ArrayList<Jugador> todos)
    {
        
        
        
        if(jugadorCorrecto(iactual, todos)){
            
            informe(iactual, todos);
            todos.get(iactual).pagaImpuesto(importe);
            
        }
                 
    }
    
    private void recibeJugador_juez(int iactual , ArrayList<Jugador> todos)
    {
        
        if(jugadorCorrecto(iactual, todos))
        {
            informe(iactual, todos);
            todos.get(iactual).encarcelar(carcel);
        }     
    }
    
    private void recibeJugador_sorpresa(int iactual , ArrayList<Jugador> todos)
    {  
        if(jugadorCorrecto(iactual,todos))
        {
            sorpresa=mazo.siguiente();//IMPY
            informe(iactual,todos);
            sorpresa.aplicarAJugador(iactual, todos);
        }
    }
    
    
    
    
    //MAIN DE PRUEBA
    public static void main(String[]args)
    {
        int i=1;
        TituloPropiedad X=new TituloPropiedad("Calle",100*i,3,50*i,500+100*i,200);
        Casilla c = new Casilla(X);
        
        System.out.println(c.toString());
        
    }
    
    
}
